File: solutions/367. Valid Perfect Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        ## use built-in function
        # n = sqrt(num)
        # return floor(n) == n
        
        ## use binary search
        def bsearch(a, b, target):
            if a > b:
                return False
            mid = a + (b-a) // 2
            if mid * mid == target:
                return True
            if mid * mid > target:
                return bsearch(a, mid - 1, target)
            return bsearch(mid + 1, b, target)
        
        return bsearch(0,num, num)

File: solutions/367. Valid Perfect Square.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        // sqrt(2^31 - 1) = 46340.9
        int l = 1, r = min(num,46340);
        while (l <= r) {
            int mid = l + (r-l) / 2;
            int midSq = mid * mid;
            if (midSq == num)
                return true;
            else if (midSq < num)
                l = mid+1;
            else
                r = mid-1;
        }
        return false;
    }
};
